lib/scholar/cluster/optics.ex

```diff
@@ -69,14 +95,46 @@ defmodule Scholar.Cluster.OPTICS do
         s64[6]
         [0, 0, 0, 1, 1, -1]
       >
+      iex> Scholar.Cluster.OPTICS.fit(x, max_eps: 2, min_samples: 1, algorithm: :kd_tree, metric: {:minkowski, 1})
+      #Nx.Tensor<
+        s64[6]
+        [0, 1, 1, 2, 2, 3]
+      >
   """
 
   deftransform fit(x, opts \\ []) do
-    fit_p(x, NimbleOptions.validate!(opts, @opts_schema))
+    if Nx.rank(x) != 2 do
+      raise ArgumentError,
+            """
+            expected x to have shape {num_samples, num_features}, \
+            got tensor with shape: #{inspect(Nx.shape(x))}
+            """
+    end
+    {opts, algorithm_opts} = Keyword.split(opts, [:min_samples, :max_eps, :eps, :algorithm])
+    opts = NimbleOptions.validate!(opts, @opts_schema)
+    algorithm_opts = Keyword.put(algorithm_opts, :num_neighbors, opts[:min_samples])
+
+    algorithm_module =
+      case opts[:algorithm] do
+        :brute ->
+          Scholar.Neighbors.BruteKNN
+
+        :kd_tree ->
+          Scholar.Neighbors.KDTree
+
+        :random_projection_forest ->
+          Scholar.Neighbors.RandomProjectionForest
+
+        module when is_atom(module) ->
+          module
+      end
+    model = algorithm_module.fit(x, algorithm_opts)
+    {_neighbors, distances} = algorithm_module.predict(model, x)
```

Comment:
Can this logic be moved to a `defn` or `defnp`? It would be nice to have it in there, because then all of the fit/predict code would folded and jitted into one too.

Reply:
I tried but encountered an issue at lines 131 and 132:
```
model = algorithm_module.fit(x, algorithm_opts).
{_neighbors, distances} = algorithm_module.predict(model, x)
```
These lines cannot be used within a numerical function, as it results in an error:

_Invalid numerical expression: algorithm_module.fit(x, algorithm_opts)._

It seems that a similar issue exists in the module you mentioned, `Scholar.Neighbors.KNNClassifier`, where this code is also placed within a deftransform function.

Reply:
I see, we need to do this slightly different, I will write a comment above.